### PR TITLE
Criterion package

### PR DESCRIPTION
**Pull request by BrunoChevalier**
###### Motivation for this change
I wanted to use Criterion as my unit testing framework, but it wasn't present yet as a nixos package.
These are the first packages I've added, so it's best if someone thoroughly looks at this request.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/boxfort/default.nix [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{stdenv, fetchFromGitHub, cmake} :
+
+stdenv.mkDerivation {
+  name = "boxfort-0.0.1";
+  src = fetchFromGitHub {
+    owner = "diacritic";
+    repo = "BoxFort";
+    rev = "4bac60bf13a49eadb76f55343164cbe8199b8d66";
+    sha256 = "1y3iywqhplffrmmydplkc9641ndyp7x8wq9h7ap5083s6ff5nk1g";
+    fetchSubmodules = true;
```
Comment (symphorien) [inline pkgs/development/libraries/boxfort/default.nix]: Is it really necessary ? I haven't seen any submodule on github.

File: pkgs/development/libraries/boxfort/default.nix [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  };
+
+  buildInputs = [ cmake ];
+
+  meta = {
+    homepage = "https://github.com/diacritic/BoxFort";
+    description = "A simple, cross-platform sandboxing C library powering Criterion.";
```
Comment (symphorien) [inline pkgs/development/libraries/boxfort/default.nix]: The description should not end with a period: see https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md

File: pkgs/development/libraries/criterion/default.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  src = fetchFromGitHub {
+    owner = "Snaipe";
+    repo = "Criterion";
+    rev = "9b70365825aced7333d7867bb5c64c63919ce510";
+    sha256 = "1i1ccgrnsq8ka5a0hfipd48baya138m2v411myg6jqkhwknplzz9";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ pkgconfig ];
+  buildInputs = [ cmake dyncall boxfort csptr nanomsg libgit2 ];
```
Comment (symphorien) [inline pkgs/development/libraries/criterion/default.nix]: `cmake` should be a `nativeBuildInput`

File: pkgs/development/libraries/csptr/default.nix [hunk 4]
```diff
@@ -0,0 +1,22 @@
+{stdenv, fetchFromGitHub, cmake} :
+
+stdenv.mkDerivation {
+  name = "csptr-2.0.4";
+  src = fetchFromGitHub {
+    owner = "Snaipe";
+    repo = "libcsptr";
+    rev = "82d3b1beafe7ed9f13f75e0d3367c60205e10f27";
+    sha256 = "0i1498h2i6zq3fn3zf3iw7glv6brn597165hnibgwccqa8sh3ich";
+    fetchSubmodules = true;
```
Comment (symphorien) [inline pkgs/development/libraries/csptr/default.nix]: There does not seem to be any submodule to fetch.

File: pkgs/development/libraries/csptr/default.nix [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  };
+
+  buildInputs = [ cmake ];
```
Comment (symphorien) [inline pkgs/development/libraries/csptr/default.nix]: Should be a `nativeBuildInput`

File: pkgs/development/libraries/csptr/default.nix [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+  meta = {
+    homepage = "https://github.com/Snaipe/libcsptr";
+    description = "This project is an attempt to bring smart pointer constructs to the (GNU) C programming language.";
```
Comment (symphorien) [inline pkgs/development/libraries/csptr/default.nix]: The description should not end with a period.

File: pkgs/development/libraries/boxfort/default.nix (hunk 1 above)
Comment (BrunoChevalier) [reply pkgs/development/libraries/boxfort/default.nix]: True, I'll remove it

**Comment by BrunoChevalier** [comment]
I guess I'll also need to reword the commit messages to comply with the CONTRIBUTING.md file.
I'll close this request and will issue a new one with modified changesets.

**Comment by symphorien** [comment]
it is enough to force-push your rephrased commits.

**Comment by symphorien** [comment]
Also, have tou tried setting `doCheck=true` on any of these packages ?

**Comment by BrunoChevalier** [comment]
No, not yet. I'll go over the complete checklist to see what I can do.

**Comment by BrunoChevalier** [comment]
Ok, doCheck=true fails for boxfort. I'm trying to figure out why the tests fail.
The other packages don't have a "make check" target.

**Comment by symphorien** [comment]
I'm think the policy for unversioned packages is to use the date of the last commit as a version: https://nixos.org/nixpkgs/manual/#sec-package-naming
Apart from this, LGTM.

File: pkgs/development/libraries/criterion/default.nix [hunk 7]
```diff
@@ -0,0 +1,25 @@
+{stdenv, pkgconfig, fetchFromGitHub, cmake, dyncall, nanomsg, boxfort, csptr, libgit2} :
+
+stdenv.mkDerivation {
+  name = "criterion-2.3.2";
```
Comment (ryantm) [inline pkgs/development/libraries/criterion/default.nix]:
```suggestion
  pname = "criterion";
  version = "2.3.2";
```

File: pkgs/development/libraries/criterion/default.nix [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  src = fetchFromGitHub {
+    owner = "Snaipe";
+    repo = "Criterion";
+    rev = "9b70365825aced7333d7867bb5c64c63919ce510";
```
Comment (ryantm) [inline pkgs/development/libraries/criterion/default.nix]:
```suggestion
    rev = "v${version}";
```

File: pkgs/development/libraries/criterion/default.nix (hunk 7 above)
Comment (ryantm) [reply pkgs/development/libraries/criterion/default.nix]: FYI, 2.3.3 is out now. It's not a requirement that you have it at the latest version to get this merged though.

File: pkgs/development/libraries/csptr/default.nix [hunk 9]
```diff
@@ -0,0 +1,21 @@
+{stdenv, fetchFromGitHub, cmake} :
+
+stdenv.mkDerivation {
+  name = "csptr-2.0.4";
```
Comment (ryantm) [inline pkgs/development/libraries/csptr/default.nix]:
```suggestion
  pname = "csptr";
  version = "2.0.4";
```

File: pkgs/development/libraries/csptr/default.nix [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  src = fetchFromGitHub {
+    owner = "Snaipe";
+    repo = "libcsptr";
+    rev = "82d3b1beafe7ed9f13f75e0d3367c60205e10f27";
```
Comment (ryantm) [inline pkgs/development/libraries/csptr/default.nix]:
```suggestion
    rev = "v${version}";
```

File: pkgs/development/libraries/boxfort/default.nix [hunk 11]
```diff
@@ -0,0 +1,21 @@
+{stdenv, fetchFromGitHub, cmake} :
+
+stdenv.mkDerivation {
+  name = "boxfort-0.0.1";
```
Comment (ryantm) [inline pkgs/development/libraries/boxfort/default.nix]:
```suggestion
  pname = "boxfort";
  version = "unstable-2018-05-10";
```

**Review by ryantm — CHANGES_REQUESTED** [review]
I'm requesting these changes so that versions can be updated more easily. Boxfort needs to be unstable because as far as I can tell 0.0.1 is not a valid upstream version.

Please also consider adding yourself as a maintainer of these packages.

**Comment by mmahut** [comment]
Are there any updates on this pull request, please?

**Comment by Thesola10** [comment]
I'm working on a derivation (#70609) and wanted to tell you -- your `csptr` derivation is unnecessary, since `libcsptr` already exists.

**Comment by Thesola10** [comment]
I am working on a new derivation (#70609) and wanted to tell you, your `csptr` derivation looks unnecessary, since `libcsptr` already exists

**Comment by BrunoChevalier** [comment]
I forgot about this pull request.
This is obsolete at this point in time as @Thesola10 and @Yumasi have added and are maintaining working derivations for Criterion and Boxfort.
I stepped away of using NixOS as my OS in the meantime because it's too rigid for me to get things done quickly.
I love the ideas but I needed more flexibility and practicality.
In a couple of years I'll probably revisit NixOS.

**Pull request closed without merging**